### PR TITLE
fix: in conditionals, don't render anything after an else branch

### PR DESCRIPTION
**Pull request by joel-hamilton**
This PR enforces a simple rule: don't render anything after an `else` branch.

There are several bugs currently where `if`, `unless` and `case` conditionals can have branches that occur after an `else` render, or that branches after an `else` can actually cause the `else` branch itself not to render.

Fixes: #670 and #672

## Example
```liquid
<p>If</p>
{% if false %}
  don't show
  {%- else -%}
  show
  {%- else -%}
  don't show
{% endif %}

<p>If with Elsif</p>
{% if false %}
  don't show
  {%- else -%}
  show
  {%- elsif true -%}
  don't show
{% endif %}

<p>Unless</p>
{% unless true %}
  don't show
  {%- else -%}
  show
  {%- else -%}
  don't show
{% endunless %}

<p>Case</p>
{% case true %}
  {%- when false- %}
    don't show
  {%- else -%}
    show
  {%- else -%}
    don't show
{% endcase %}

<p>Case with When</p>
{% case true -%}
  {% when false -%}
    don't show (false)
  {%- else -%}
    show
  {%- else -%}
    don't show (second else)
  {%- when true -%}
    don't show (truthy when condition after else)
  {%- when true -%}
    don't show (truthy when condition after else)
{%- endcase %}
```

**`master` Behaviour**:
```html
<p>If</p>
showdon't show

<p>If with Elsif</p>
don't show

<p>Unless</p>
showdon't show

<p>Case</p>
showdon't show

<p>Case with When</p>
don't show (truthy when condition after else)don't show (truthy when condition after else)
```

**PR Behaviour**:
```html
<p>If</p>
show

<p>If with Elsif</p>
show

<p>Unless</p>
show

<p>Case</p>
show
 
<p>Case with When</p>
show
```

### CODE REVIEW

File: src/tags/if.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ export default class extends Tag {
 
   constructor (tagToken: TagToken, remainTokens: TopLevelToken[], liquid: Liquid) {
     super(tagToken, remainTokens, liquid)
-    let p
+    let p: Template[] = []
```
Comment (harttle) [inline src/tags/if.ts:9]: Good job

File: src/tags/case.ts [hunk 2]
```diff
@@ -31,7 +31,11 @@ export default class extends Tag {
       })
       .on('tag:else', () => (p = this.elseTemplates))
       .on('tag:endcase', () => stream.stop())
-      .on('template', (tpl: Template) => p.push(tpl))
+      .on('template', (tpl: Template) => {
+        if (p !== this.elseTemplates || (p === this.elseTemplates && p.length === 0)) {
```
Comment (harttle) [inline src/tags/case.ts:35]: This won't work when there are multiple templates:

```
{%else%}
B{{output}}A
```

File: src/tags/case.ts [hunk 3]
```diff
@@ -31,7 +31,11 @@ export default class extends Tag {
       })
       .on('tag:else', () => (p = this.elseTemplates))
```
Comment (harttle) [inline src/tags/case.ts:32]: I think we can set p to [] which serves as a blackhole when `elaeTemplates` is already set. And we need initialize it as null instead of empty string. For case

```
{%else%}{%else%}we don't want this output
```

File: test/e2e/issues.spec.ts [hunk 4]
```diff
@@ -469,4 +469,19 @@ describe('Issues', function () {
     const result = engine.parseAndRenderSync('{{ÜLKE}}', { ÜLKE: 'Türkiye' })
     expect(result).toEqual('Türkiye')
   })
+  it('#670 Does not render multiple `else` branches in the same if conditional', () => {
```
Comment (harttle) [inline test/e2e/issues.spec.ts:472]: We can move edge cases into tags/, while preserving the case mentioned in that issue.

**Review by harttle — CHANGES_REQUESTED** [review]

File: src/tags/case.ts (hunk 3 above)
Comment (joel-hamilton) [reply src/tags/case.ts]: I used the `p = []` and an `else` counter to prevent rendering any branches that follow the `else`.

> `{%else%}{%else%}we don't want this output`

I added an empty else block to the [tests](https://github.com/joel-hamilton/liquidjs/blob/e2b22d0187916a8d31ad7772248a4668f85424d0/test/integration/tags/unless.spec.ts#L60) and it passed without changing the initialization. Let me know if you see another edge case I'm missing though.

File: src/tags/case.ts [hunk 5]
```diff
@@ -29,9 +34,16 @@ export default class extends Tag {
...
       })
-      .on('tag:else', () => (p = this.elseTemplates))
+      .on('tag:else', () => {
+        elseCount++
+        p = this.elseTemplates
+      })
       .on('tag:endcase', () => stream.stop())
-      .on('template', (tpl: Template) => p.push(tpl))
+      .on('template', (tpl: Template) => {
+        if (p !== this.elseTemplates || (p === this.elseTemplates && elseCount === 1)) {
```
Comment (harttle) [inline src/tags/case.ts:43]:
```suggestion
        if (p !== this.elseTemplates || elseCount === 1) {
```

File: src/tags/if.ts [hunk 6]
```diff
@@ -6,19 +6,33 @@ export default class extends Tag {
...
+        })
+      })
+      .on('tag:else', () => {
+        elseCount++
+        p = this.elseTemplates
+      })
       .on('tag:endif', function () { this.stop() })
-      .on('template', (tpl: Template) => p.push(tpl))
+      .on('template', (tpl: Template) => {
+        if (p !== this.elseTemplates || (p === this.elseTemplates && elseCount === 1)) {
```
Comment (harttle) [inline src/tags/if.ts:32]:
```suggestion
        if (p !== this.elseTemplates || elseCount === 1) {
```

File: src/tags/unless.ts [hunk 7]
```diff
@@ -5,21 +5,35 @@ export default class extends Tag {
...
+        })
+      })
+      .on('tag:else', () => {
+        elseCount++
+        p = this.elseTemplates
+      })
       .on('tag:endunless', function () { this.stop() })
-      .on('template', (tpl: Template) => p.push(tpl))
+      .on('template', (tpl: Template) => {
+        if (p !== this.elseTemplates || (p === this.elseTemplates && elseCount === 1)) {
```
Comment (harttle) [inline src/tags/unless.ts:33]:
```suggestion
        if (p !== this.elseTemplates || elseCount === 1) {
```

**Comment by harttle** [comment]
@all-contributors please add @joel-hamilton for code
